File: SDK Examples/Examples/VB.Net/Counter_BufferedPulseWidthMeasurement/BufferedPulseWidthMeasurementForm.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Collections
Imports System.Drawing.Drawing2D
Imports Automation.BDaq

Partial Public Class BufferedPulseWidthMeasurementForm
   Inherits Form
   Protected Delegate Sub UpdateListview(ByVal eventChannel As Integer, ByVal displayChannel As Integer)
   Protected Delegate Sub StoppedEvent()
   Private m_isRunning As Boolean = False
   Private m_isCountingReset As Boolean = True
   Private m_flag As Boolean = False
   Const m_maxOfCounterValueList As Integer = 10 ' The max value of lines of CounterValueList.
   Private m_dataBuf As PulseWidth()
   Private chs_dataBuf As PulseWidth()()
   Private ch_status As String()
   Private m_eventChannel As Integer
   Private m_samplesCount As Integer
   Private m_sectionLength As Integer
   Private m_channelCountMax As Integer
   Private m_displayChannel As Integer
   Private m_channelEnable As Byte()

   Public Sub New()
      InitializeComponent()
   End Sub

   Private Sub HandleError(ByVal err As ErrorCode)
      If err <> ErrorCode.Success Then
         MessageBox.Show("Sorry ! There are some errors happened, the error code is: " & err.ToString(), "BufferedPulseWidthMeasurement")
      End If
   End Sub

   Private Shared Sub ShowErrorMessage(ByVal e As Exception)
      Dim errorInfo As String
      errorInfo = "There's some error happened, the error information: "
      MessageBox.Show(errorInfo & e.Message)
   End Sub

   Private Sub BufferedPulseWidthMeasurementrForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
      'no device is selected
      If Not BufferedPwMeterCtrl1.Initialized Then
         MessageBox.Show("Please select a device in control property!")
         Me.Close()
         Return
      End If

      AddHandler cbxEventChannel.SelectedIndexChanged, AddressOf onCbxEventChannelSelectedIndexChanged

      m_channelCountMax = BufferedPwMeterCtrl1.Features.ChannelCountMax
      m_channelEnable = New Byte(m_channelCountMax) {}
      m_channelEnable = BufferedPwMeterCtrl1.ChannelMap

      m_sectionLength = BufferedPwMeterCtrl1.Record.SectionLength
      Dim sectionCount As Integer = BufferedPwMeterCtrl1.Record.SectionCount
      m_samplesCount = m_sectionLength
      If sectionCount Then
         m_samplesCount = sectionCount * m_samplesCount
      End If
      m_dataBuf = New PulseWidth(m_samplesCount) {}
      chs_dataBuf = New PulseWidth(m_channelCountMax)() {}
      Dim i As Integer = 0
      For i = 0 To m_channelCountMax - 1
         chs_dataBuf(i) = New PulseWidth(m_samplesCount) {}
      Next

      'initial all channels status are Ready.
      ch_status = New String(m_channelCountMax) {}

      'configure combox;
      cbxEventChannel.BeginUpdate()
      For i = 0 To m_channelCountMax - 1
         If m_channelEnable(i) = 1 Then
            cbxEventChannel.Items.Add(i.ToString)
         End If
      Next
      cbxEventChannel.EndUpdate()
      cbxEventChannel.SelectedIndex = 0

      'display the status of default channel .
      For i = 0 To m_channelCountMax - 1
         If m_channelEnable(i) = 1 Then
            ChannelStatusStringChange(i, "Ready")
         End If
      Next
      Dim str As String = cbxEventChannel.Text
      m_displayChannel = Convert.ToInt32(str)
      DisplayChannelStatus(m_displayChannel)

      'set title of the form.
      Dim description As String = BufferedPwMeterCtrl1.SelectedDevice.Description
      Me.Text = "Buffered Pulse Width Measurement - Run(" & description & ")"
      CounterValueList.Items.Clear()
      btnStart.Enabled = True
      btnStop.Enabled = False
   End Sub

   Protected Sub UpdateListviewMethod(ByVal eventChannel As Integer, ByVal displayChannel As Integer)
      CopyChannelData(eventChannel)
      If eventChannel = displayChannel Then
         DisplayChannelValue(displayChannel)
      End If
   End Sub

   Protected Sub StoppedEventMethod()
      Try
         If BufferedPwMeterCtrl1.State = ControlState.Stopped Then
            BufferedPwMeterCtrl1.Enabled = False
            m_isRunning = False
            m_isCountingReset = True
            btnStart.Enabled = True
            btnStop.Enabled = False
         End If
      Catch ex As Exception
         ShowErrorMessage(ex)
      End Try
      ChannelStatusStringChange(m_eventChannel, "Stopped")

      DisplayChannelStatus(m_displayChannel)
   End Sub

   Private Sub Start_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnStart.Click
      If m_isCountingReset = True Then
         CounterValueList.Items.Clear()
         Dim i As Int32 = 0
         For i = 0 To m_channelCountMax - 1
            If m_channelEnable(i) = 1 Then
               ChannelStatusStringChange(i, "Running")
            End If
         Next
         DisplayChannelStatus(m_displayChannel)

         Try
            Dim err As ErrorCode = ErrorCode.Success
            err = BufferedPwMeterCtrl1.Prepare()
            If err <> ErrorCode.Success Then
               HandleError(err)
               Return
            End If

            BufferedPwMeterCtrl1.Enabled = True
            m_isRunning = True

            btnStop.Enabled = True
            btnStart.Enabled = False
            m_isCountingReset = False
            m_flag = True
         Catch ex As Exception
            ShowErrorMessage(ex)
         End Try
      End If
   End Sub

   Private Sub Stop_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnStop.Click
      'Stop event counting
      Try
         BufferedPwMeterCtrl1.Enabled = False
         m_isRunning = False
      Catch ex As Exception
         ShowErrorMessage(ex)
      End Try

      m_isCountingReset = True
      btnStart.Enabled = True
      btnStop.Enabled = False

      Dim i As Int32 = 0
      For i = 0 To m_channelCountMax - 1
         If m_channelEnable(i) = 1 Then
            ChannelStatusStringChange(i, "Stopped")
         End If
      Next
      DisplayChannelStatus(m_displayChannel)
   End Sub
   
   Private Sub BufferedPwMeterCtrl1_DataReady(ByVal sender As System.Object, ByVal args As Automation.BDaq.BfdCntrEventArgs) Handles BufferedPwMeterCtrl1.DataReady
      Dim err As ErrorCode = ErrorCode.Success
      Dim getDataCount As Integer = Math.Min(m_samplesCount, args.Count)
      Dim returnedCount As Integer = 0
      Dim startTime As Double = 0.0
      m_eventChannel = args.Channel
      err = BufferedPwMeterCtrl1.GetData(args.Channel, getDataCount, m_dataBuf, 0, returnedCount, startTime)
      If err <> ErrorCode.Success Then
         HandleError(err)
         Return
      End If
      Try
         Me.Invoke(New UpdateListview(AddressOf UpdateListviewMethod), New Object() {m_eventChannel, m_displayChannel})
      Catch ex As System.Exception
      End Try
   End Sub

   Private Sub BufferedPwMeterCtrl1_StoppedEvent(ByVal sender As System.Object, ByVal args As Automation.BDaq.BfdCntrEventArgs) Handles BufferedPwMeterCtrl1.Stopped
      Dim err As ErrorCode = ErrorCode.Success
      Dim getDataCount As Integer = Math.Min(m_samplesCount, args.Count)
      Dim returnedCount As Integer = 0
      Dim startTime As Double = 0.0
      m_eventChannel = args.Channel
      If getDataCount > 0 Then
         err = BufferedPwMeterCtrl1.GetData(args.Channel, getDataCount, m_dataBuf, 0, returnedCount, startTime)
         If err <> ErrorCode.Success Then
            HandleError(err)
            Return
         End If

         Try
            Me.Invoke(New UpdateListview(AddressOf UpdateListviewMethod), New Object() {m_eventChannel, m_displayChannel})
         Catch ex As System.Exception
         End Try
      End If

      Try
         Me.Invoke(New StoppedEvent(AddressOf StoppedEventMethod), New Object() {})
      Catch ex As Exception
      End Try
   End Sub

   Private Sub onCbxEventChannelSelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs)
      m_displayChannel = Convert.ToInt32(cbxEventChannel.Text)
      DisplayChannelValue(m_displayChannel)
      DisplayChannelStatus(m_displayChannel)
   End Sub

   Private Sub ChannelStatusStringChange(ByVal channel As Integer, ByVal status As String)
      Dim str As String = status
      Dim ch As Integer = channel
      ch_status(ch) = str
   End Sub

   Private Sub DisplayChannelStatus(ByVal channel As Integer)
      channelStatus.Text = ch_status(channel)
   End Sub

   Private Sub DisplayChannelValue(ByVal channel As Integer)
      Dim i As Int32 = 0
      CounterValueList.Items.Clear()
      Dim itemCount As Integer = 0
      If m_sectionLength >= 10 Then
         itemCount = 10
      ElseIf m_sectionLength < 10 Then
         itemCount = m_sectionLength
      End If

      Try
         DisplayChannelStatus(channel)
         If m_isRunning Or m_flag Then
            For i = 0 To itemCount - 1
               CounterValueList.Items.Add("High:" + chs_dataBuf(channel)(i).HiPeriod.ToString("0.000000") + " s Low:" + chs_dataBuf(channel)(i).LoPeriod.ToString("0.000000") + " s")
            Next
         End If
      Catch ex As Exception
         ShowErrorMessage(ex)
      End Try
   End Sub

   Private Sub CopyChannelData(ByVal channel As Integer)
      Dim ch As Integer = channel
      Array.Copy(m_dataBuf, chs_dataBuf(ch), m_dataBuf.Length)
   End Sub
End Class

File: SDK Examples/Examples/C#/Counter_BufferedPulseWidthMeasurement/BufferedPulseWidthMeasurementForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Collections;
using System.Drawing.Drawing2D;
using Automation.BDaq;


namespace Counter_BufferedPulseWidthMeasurement
{
   public partial class BufferedPulseWidthMeasurementForm : Form
   {
      protected delegate void UpdateListview(int eventChannel, int displayChannel);
      protected delegate void StopEvent();

      bool m_isRunning = false;
      bool m_isCountingReset = true;
      bool m_flag = false;
      const int m_maxOfCounterValueList = 10; // The max value of lines of CounterValueList.
      PulseWidth[] m_dataBuf;
      PulseWidth[][] chs_dataBuf;
      string[] ch_status;
      int m_eventChannel;
      int m_samplesCount;
      int m_sectionLength;
      int m_channelCountMax;
      int m_displayChannel;
      byte[] m_channelEnable;

      public BufferedPulseWidthMeasurementForm()
      {
         InitializeComponent();
      }

      private void HandleError(ErrorCode err)
      {
         if (err != ErrorCode.Success)
         {
            MessageBox.Show("Sorry ! Some errors happened, the error code is: " + err.ToString(), "BufferedPulseWidthMeasurement");
         }
      }

      private static void ShowErrorMessage(Exception e)
      {
         string errorInfo;
         errorInfo = "There's some error happened, the error information: ";
         MessageBox.Show(errorInfo + e.Message);
      }

      private void BufferedPulseWidthMeasurementForm_Load(object sender, EventArgs e)
      {
         //The default device of project is demo device, users can choose other devices according to their needs. 
         //// no device is selected
         if (!bufferedPwMeterCtrl1.Initialized)
         {
            MessageBox.Show("No device be selected or device open failed!");
            this.Close();
            return;
         }

         cbxEventChannel.SelectedIndexChanged += new EventHandler(cbxEventChannel_SelectedIndexChanged);

         m_channelCountMax = bufferedPwMeterCtrl1.Features.ChannelCountMax;
         m_channelEnable = new byte[m_channelCountMax];
         m_channelEnable = bufferedPwMeterCtrl1.ChannelMap;
         m_sectionLength = bufferedPwMeterCtrl1.Record.SectionLength;
         int sectionCount = bufferedPwMeterCtrl1.Record.SectionCount;
         m_samplesCount = m_sectionLength;
         if (sectionCount != 0)
         {
            m_samplesCount *= sectionCount;
         }
         m_dataBuf = new PulseWidth[m_samplesCount];

         chs_dataBuf = new PulseWidth[m_channelCountMax][];
         for (int i = 0; i < m_channelCountMax; i++)
         {
            chs_dataBuf[i] = new PulseWidth[m_samplesCount];
         } 
         //initial all channels status are Ready.
         ch_status = new string[m_channelCountMax];

         //configure combobox;
         for (int i = 0; i < m_channelCountMax; i++)
         {
            if (m_channelEnable[i] == 1)
            {
               cbxEventChannel.Items.Add(i);
            }
         }
         cbxEventChannel.SelectedIndex = 0;

         // display the status of default channel.
         for (int i = 0; i < m_channelCountMax; i++)
         {
            if (m_channelEnable[i] == 1)
            {
               ChannelStatusStringChange(i, "Ready");
            }
         }
         m_displayChannel = Convert.ToInt32(cbxEventChannel.Text);
         DisplayChannelStatus(m_displayChannel);

         //set title of the form.
         string description = bufferedPwMeterCtrl1.SelectedDevice.Description;
         this.Text = "Buffered Pulse Width Measurement(" + description + ")";

         CounterValueList.Items.Clear();
         Start.Enabled = true;
         Stop.Enabled = false;
      }

      private void UpdateListviewMethod(int eventChannel, int displayChannel)
      {
         CopyChannelData(eventChannel);
         if (eventChannel == displayChannel){
            DisplayChannelValue(displayChannel);
         }
      }

      private void StopEventMethod()
      {
         try
         {
            if (bufferedPwMeterCtrl1.State == ControlState.Stopped)
            {
               bufferedPwMeterCtrl1.Enabled = false;
               m_isRunning = false;
               m_isCountingReset = true;
               Start.Enabled = true;
               Stop.Enabled = false;
            }
         }
         catch (Exception ex)
         {
            ShowErrorMessage(ex);
         }

         ChannelStatusStringChange(m_eventChannel, "Stopped");
         DisplayChannelStatus(m_displayChannel);
      }

      private void Start_Click(object sender, EventArgs e)
      {
         ErrorCode err = ErrorCode.Success;
         if (m_isCountingReset == true)
         {
            CounterValueList.Items.Clear();
            for (int i = 0; i < m_channelCountMax; i++)
            {
               if (m_channelEnable[i] == 1)
               {
                  ChannelStatusStringChange(i, "Running");
               }
            }
            DisplayChannelStatus(m_displayChannel);

            err = bufferedPwMeterCtrl1.Prepare();
            if (err != ErrorCode.Success)
            {
               HandleError(err);
               return;
            }
            bufferedPwMeterCtrl1.Enabled = true;
            m_isRunning = true;
            Start.Enabled = false;
            Stop.Enabled = true;
            m_isCountingReset = false;
            m_flag = true;
         }
      }

      private void Stop_Click(object sender, EventArgs e)
      {
         // Stop event counting
         try
         {
            bufferedPwMeterCtrl1.Enabled = false;
            m_isRunning = false;
         }
         catch (Exception ex)
         {
            ShowErrorMessage(ex);
         }

         m_isCountingReset = true;
         Start.Enabled = true;
         Stop.Enabled = false;
         for (int i = 0; i < m_channelCountMax; i++)
         {
            if (m_channelEnable[i] == 1)
            {
               ChannelStatusStringChange(i, "Stopped");
            }
         }
         DisplayChannelStatus(m_displayChannel);
      }

      private void bufferedPwMeterCtrl1_DataReady(object sender, BfdCntrEventArgs e)
      {
         ErrorCode ret = ErrorCode.Success;
         int getDataCount = Math.Min(m_samplesCount, e.Count);
         int returnCount = 0;
         double startTime = 0.0;
         m_eventChannel = e.Channel;
         ret = bufferedPwMeterCtrl1.GetData(e.Channel, getDataCount, m_dataBuf, 0, out returnCount, out startTime);
         if (ret != ErrorCode.Success)
         {
            HandleError(ret);
         }

         try
         {
            Invoke(new UpdateListview(UpdateListviewMethod), new object[] {m_eventChannel, m_displayChannel});
         }
         catch (System.Exception ex) { }
      }

      private void bufferedPwMeterCtrl1_StopEvent(object sender, BfdCntrEventArgs e)
      {
         ErrorCode ret = ErrorCode.Success;
         int getDataCount = Math.Min(m_samplesCount, e.Count);
         int returnCount = 0;
         double startTime = 0.0;
         m_eventChannel = e.Channel;
         if(getDataCount > 0){
            ret = bufferedPwMeterCtrl1.GetData(e.Channel, getDataCount, m_dataBuf, 0, out returnCount, out startTime);
            if (ret != ErrorCode.Success)
            {
               HandleError(ret);
            }

            try
            {
               Invoke(new UpdateListview(UpdateListviewMethod), new object[] { m_eventChannel, m_displayChannel });
            }
            catch (System.Exception ex) { }
         }
         
         try
         {
            Invoke(new StopEvent(StopEventMethod), new object[] { });
         }
         catch (System.Exception ex) { }
      }

      private void cbxEventChannel_SelectedIndexChanged(object sender, EventArgs e)
      {
         m_displayChannel = Convert.ToInt32((cbxEventChannel.Text));
         DisplayChannelValue(m_displayChannel);
         DisplayChannelStatus(m_displayChannel);
      }

      void ChannelStatusStringChange(int channel, string status)
      {
         string str = status;
         int ch = channel;
         ch_status[ch] = str;
      }

      void DisplayChannelStatus(int channel)
      {
         channelStatus.Text = ch_status[channel];
      }

      void DisplayChannelValue(int channel)
      {
         CounterValueList.Items.Clear();
         int listCount = 0;
         if (m_sectionLength >= 10){
            listCount = 10;
         }else{
            listCount = m_sectionLength;
         }

         try
         {
            if (m_isRunning | m_flag)
            {
               for (int i = 0; i < listCount; i++)
               {
                  CounterValueList.Items.Add("High:" + chs_dataBuf[channel][i].HiPeriod.ToString("0.000000") + " s Low:" + chs_dataBuf[channel][i].LoPeriod.ToString("0.000000") + " s");
               }
            }
         }
         catch (Exception ex)
         {
            ShowErrorMessage(ex);
         }
      }

      void CopyChannelData(int channel)
      {
         int ch = channel;
         Array.Copy(m_dataBuf, 0, chs_dataBuf[ch], 0, m_dataBuf.Length);
      }
   }
}